### ISSUE TITLE
Error when running ECnQSR

### ISSUE THREAD

**Issue by XU-Nuo** [mingjiewang/qap#11]
**Command:**
```
qap ECnQSR \
    --fasta ./6.ampliconSeqTrim/test1_P5.demo.fasta \
    --amplicon Y \
    --program ShoRAH \
    --threads 4 \
    --refSeq ./7.ampliconSeqCS/test1_P5.demo_CS.fasta \
    --outputDir ./8.seqEC 
```
**Result:**
```
You are now running subprogram:  ECnQSR

WARNING @ [2021-05-18 06:37:25 UTC]: The output directory /tmp/input/8.4seqEC/ does NOT exist.
WARNING @ [2021-05-18 06:37:25 UTC]: Will mkdir /tmp/input/8.4seqEC/ and use it as the output directory.
INFO    @ [2021-05-18 06:37:25 UTC]: Start to map reads to /tmp/input/7.ampliconSeqCS/test1_P5.demo_CS.fasta using Bowtie2.
INFO    @ [2021-05-18 06:37:25 UTC]: Running Bowtie2 for mapping
INFO    @ [2021-05-18 06:37:25 UTC]: /opt/qap/bin/3rdPartyTools/bowtie2/bowtie2 --very-sensitive-local -x /tmp/input/7.ampliconSeqCS/test1_P5.demo_CS -f /tmp/input/8.4seqEC/tmp/test1_P5.demo.fasta -S /tmp/input/8.4seqEC/tmp/test1_P5.demo.sam
500 reads; of these:
  500 (100.00%) were unpaired; of these:
    0 (0.00%) aligned 0 times
    500 (100.00%) aligned exactly 1 time
    0 (0.00%) aligned >1 times
100.00% overall alignment rate
INFO    @ [2021-05-18 06:37:25 UTC]: Converting /tmp/input/8.4seqEC/tmp/test1_P5.demo.sam to bam file
Use of uninitialized value $file in -e at /opt/qap/bin/../lib/General.pm line
	764 (#1)
    (W uninitialized) An undefined value was used as if it were already
    defined.  It was interpreted as a "" or a 0, but maybe it was a mistake.
    To suppress this warning assign a defined value to your variables.
    
    To help you figure out what was undefined, perl will try to tell you
    the name of the variable (if any) that was undefined.  In some cases
    it cannot do this, so it also tells you what operation you used the
    undefined value in.  Note, however, that perl optimizes your program
    and the operation displayed in the warning may not necessarily appear
    literally in your program.  For example, "that $foo" is usually
    optimized into "that " . $foo, and the warning will refer to the
    concatenation (.) operator, even though there is no . in
    your program.
    
Use of uninitialized value $bamfile_unsort in concatenation (.) or string at
	/opt/qap/bin/../lib/Mapper.pm line 291 (#1)
INFO    @ [2021-05-18 06:37:25 UTC]: /opt/qap/bin/3rdPartyTools/samtools/samtools view -h -@ 4 -Sb /tmp/input/8.4seqEC/tmp/test1_P5.demo.sam > 
sh: 1: Syntax error: end of file unexpected
Use of uninitialized value $bamfile_unsort in concatenation (.) or string at
	/opt/qap/bin/../lib/Mapper.pm line 297 (#1)
INFO    @ [2021-05-18 06:37:25 UTC]: Sorting bam file  with coordinates.
Use of uninitialized value $bamfile_unsort in concatenation (.) or string at
	/opt/qap/bin/../lib/Mapper.pm line 306 (#1)
INFO    @ [2021-05-18 06:37:25 UTC]: /opt/qap/bin/3rdPartyTools/samtools/samtools sort -@ 4 -O bam -T test1_P5.demo.PosSorted.1621319845 -o /tmp/input/8.4seqEC/tmp/test1_P5.demo.PosSorted.bam 
Usage: samtools sort [options...] [in.bam]
Options:
  -l INT     Set compression level, from 0 (uncompressed) to 9 (best)
  -m INT     Set maximum memory per thread; suffix K/M/G recognized [768M]
  -n         Sort by read name
  -t TAG     Sort by value of TAG. Uses position as secondary index (or read name if -n is set)
  -o FILE    Write final output to FILE rather than standard output
  -T PREFIX  Write temporary files to PREFIX.nnnn.bam
      --input-fmt-option OPT[=VAL]
               Specify a single input file format option in the form
               of OPTION or OPTION=VALUE
  -O, --output-fmt FORMAT[,OPT[=VAL]]...
               Specify output format (SAM, BAM, CRAM)
      --output-fmt-option OPT[=VAL]
               Specify a single output file format option in the form
               of OPTION or OPTION=VALUE
      --reference FILE
               Reference sequence FASTA FILE [null]
  -@, --threads INT
               Number of additional threads to use [0]
INFO    @ [2021-05-18 06:37:25 UTC]: Indexing sorted bam file.
INFO    @ [2021-05-18 06:37:25 UTC]: /opt/qap/bin/3rdPartyTools/samtools/samtools index /tmp/input/8.4seqEC/tmp/test1_P5.demo.PosSorted.bam
[E::hts_open_format] Failed to open file /tmp/input/8.4seqEC/tmp/test1_P5.demo.PosSorted.bam
samtools index: failed to open "/tmp/input/8.4seqEC/tmp/test1_P5.demo.PosSorted.bam": No such file or directory
Use of uninitialized value $bamfile_unsort in concatenation (.) or string at
	/opt/qap/bin/../lib/Mapper.pm line 336 (#1)
INFO    @ [2021-05-18 06:37:25 UTC]: Converting /tmp/input/8.4seqEC/tmp/test1_P5.demo.sam to bam file
INFO    @ [2021-05-18 06:37:25 UTC]: /opt/qap/bin/3rdPartyTools/samtools/samtools view -h -@ 4 -Sb /tmp/input/8.4seqEC/tmp/test1_P5.demo.sam > 
sh: 1: Syntax error: end of file unexpected
Use of uninitialized value $bamfile_unsort in concatenation (.) or string at
	/opt/qap/bin/../lib/Mapper.pm line 310 (#1)
INFO    @ [2021-05-18 06:37:25 UTC]: Sorting bam file  with reads name.
Use of uninitialized value $bamfile_unsort in concatenation (.) or string at
	/opt/qap/bin/../lib/Mapper.pm line 319 (#1)
INFO    @ [2021-05-18 06:37:25 UTC]: /opt/qap/bin/3rdPartyTools/samtools/samtools sort -n -@ 4 -O bam -T test1_P5.demo.NameSorted.1621319845 -o /tmp/input/8.4seqEC/tmp/test1_P5.demo.NameSorted.bam 
Usage: samtools sort [options...] [in.bam]
Options:
  -l INT     Set compression level, from 0 (uncompressed) to 9 (best)
  -m INT     Set maximum memory per thread; suffix K/M/G recognized [768M]
  -n         Sort by read name
  -t TAG     Sort by value of TAG. Uses position as secondary index (or read name if -n is set)
  -o FILE    Write final output to FILE rather than standard output
  -T PREFIX  Write temporary files to PREFIX.nnnn.bam
      --input-fmt-option OPT[=VAL]
               Specify a single input file format option in the form
               of OPTION or OPTION=VALUE
  -O, --output-fmt FORMAT[,OPT[=VAL]]...
               Specify output format (SAM, BAM, CRAM)
      --output-fmt-option OPT[=VAL]
               Specify a single output file format option in the form
               of OPTION or OPTION=VALUE
      --reference FILE
               Reference sequence FASTA FILE [null]
  -@, --threads INT
               Number of additional threads to use [0]
INFO    @ [2021-05-18 06:37:25 UTC]: Indexing sorted bam file.
INFO    @ [2021-05-18 06:37:25 UTC]: /tmp/input/8.4seqEC/tmp/test1_P5.demo.NameSorted.bam is sorted by name, so skipping samtools index. 
Use of uninitialized value $samfile in concatenation (.) or string at
	/opt/qap/bin/../lib/Mapper.pm line 471 (#1)
INFO    @ [2021-05-18 06:37:25 UTC]: Converting bam file /tmp/input/8.4seqEC/tmp/test1_P5.demo.NameSorted.bam to 
Use of uninitialized value $samfile in concatenation (.) or string at
	/opt/qap/bin/../lib/Mapper.pm line 472 (#1)
INFO    @ [2021-05-18 06:37:25 UTC]: /opt/qap/bin/3rdPartyTools/samtools/samtools view -h -@ 4 /tmp/input/8.4seqEC/tmp/test1_P5.demo.NameSorted.bam > 
sh: 1: Syntax error: end of file unexpected
INFO    @ [2021-05-18 06:37:25 UTC]: Running Shorah for ECnQSR
INFO    @ [2021-05-18 06:37:25 UTC]: python /opt/qap/bin/3rdPartyTools/qsr/Shorah/amplian.py -b /tmp/input/8.4seqEC/tmp/test1_P5.demo.PosSorted.bam -f /tmp/input/8.4seqEC/tmp/test1_P5.demo_CS.2line.fasta
open: No such file or directory
Failed to open BAM file /tmp/input/8.4seqEC/tmp/test1_P5.demo.PosSorted.bam
Traceback (most recent call last):
  File "/opt/conda/lib/python2.7/logging/handlers.py", line 76, in emit
    if self.shouldRollover(record):
  File "/opt/conda/lib/python2.7/logging/handlers.py", line 156, in shouldRollover
    msg = "%s\n" % self.format(record)
  File "/opt/conda/lib/python2.7/logging/__init__.py", line 741, in format
    return fmt.format(record)
  File "/opt/conda/lib/python2.7/logging/__init__.py", line 465, in format
    record.message = record.getMessage()
  File "/opt/conda/lib/python2.7/logging/__init__.py", line 329, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Logged from file amplian.py, line 61
Traceback (most recent call last):
  File "/opt/qap/bin/3rdPartyTools/qsr/Shorah/amplian.py", line 437, in <module>
    main(*args, **vars(options))
  File "/opt/qap/bin/3rdPartyTools/qsr/Shorah/amplian.py", line 341, in main
    h = list(open('coverage.txt'))[0]
IOError: [Errno 2] No such file or directory: 'coverage.txt'
INFO    @ [2021-05-18 06:37:26 UTC]: Program completed!
```

**Comment by XU-Nuo** [comment]
This run time failure consists of 3 parts:
- The `Use of uninitialized value $file` in issue #12.
- The `samtool` is not correctly installed in issue #13.
- The `shorah` is not correctly installed in issue #14.

**Comment by XU-Nuo** [comment]
Closed together with #14
